src/user.ts

```diff
@@ -485,4 +486,35 @@ export class Session {
 		}
 		return pass;
 	}
+
+	async createPasswordResetToken(name: string, timeout: null | number = null) {
+		const ctime = time();
+		const userid = toID(name);
+		if (!timeout) {
+			timeout = 7 * 24 * 60 * 60;
+		}
+		timeout += ctime;
+		// todo throttle by checking to see if pending token exists in sid table?
+		if (await this.findPendingReset(name)) {
+			throw new ActionError(`A reset token is already pending to that account.`);
+		}
+
+		await usermodlog.insert({
+			userid, actorid: userid, ip: this.context.getIp(),
+			date: ctime, entry: "Password reset token requested",
+		});
+
+		// magical character string...
+		const token = crypto.randomBytes(PASSWORD_RESET_TOKEN_SIZE).toString('hex');
+		await sessions.insert({
+			userid, sid: token, time: ctime, timeout, ip: this.context.getIp(),
+		});
+		return token;
+	}
+	async findPendingReset(name: string) {
+		const id = toID(name);
+		const sids = await sessions.selectAll()`WHERE userid = ${id}`;
+		// not a fan of this but sids are normally different lengths. have to be, iirc.
+		return sids.some(({sid}) => sid.length === (PASSWORD_RESET_TOKEN_SIZE * 2));
+	}
```

Comment:
Can you explain what's going on here? PHP password reset tokens are 15 bytes, while yours are 10, so I'm not sure this would find manually-generated password reset tokens.

Reply:
`crypto.randomBytes(num).toString('hex')` returns a string with a length of `num * 2`, I checked.

Reply:
PHP password reset tokens should not be found by this, it was specific design. My thoughts were 1) why are you using an email reset and a manual reset request at the same time? Should that ever really be something we should do? and 2) What if they lose the URL somehow? That shouldn't block an email reset imo.

Reply:
OK, that's fine.